File: AI/robots/robot.js
'use strict';

module.exports = class Robot {
    constructor(app){
        this.app = app;
    }

    run(action) {
        if(action.method in this){ return this[action.method](action.parameters);}
        else{
            this.app.logger.log(action.method+" method not declared")
            return false;
        }
    }

    moveToElement(parameters){
        this.app.logger.log("--> Moving to "+parameters.element+"at speed "+parameters.speed);
        return true;
    }
}

File: pythonAI/robots/robot.py
class Robot():

    def __init__(self, app):
        self.app = app

    def method_exists(self, method):
        if(method in dir(self) and callable(func := getattr(self, method))):
            return (True, func)
        else:
            return (False,)

    def run(self, action):
        cond, method = self.method_exists(action['method'])
        if(cond):
            return method(action['parameters'])
        else:
            self.app['logger'].info(action['method'] + " method not declared")
            return False
    
    def moveToElement(self, parameters):
        self.app['logger'].info("--> Moving to " + str(parameters['element']) +\
                                " at speed " + str(parameters['speed']))
        return True
